File: week6/pset6/sentimental-readability/readability.py
from cs50 import get_string


def main():
    # Get the text from the user
    text = get_text()

    # Count how many letters are in the given text using count_letters() function
    letters = count_letters(text)

    # Count how many words are in the given text using count_words() function
    words = count_words(text)

    # Count how many sentences are in the given text using count_sentences() function
    sentences = count_sentences(text)

    # Calculate average number of letters per 100 words in the text using calculate_l() function
    l = calculate_l(letters, words)

    # Calculate average number of sentences per 100 words in the text using calculate_s() function
    s = calculate_s(sentences, words)

    # Calculate The Coleman-Liau index and round it to nearest integer
    index = index = 0.0588 * l - 0.296 * s - 15.8
    index = round(index)

    # Output the grade
    if index >= 16:
        print("Grade 16+")
    elif index < 1:
        print("Before Grade 1")
    else:
        print(f"Grade {index}")


# Calculates average number of sentences per 100 words in the text
def calculate_s(sentences, words):
    s = sentences / words * 100
    s = round(s, 2)
    return s


# Calculates average number of letters per 100 words in the text
def calculate_l(letters, words):
    l = letters / words * 100
    l = round(l, 2)
    return l


# Counts how many sentences are in the given text
def count_sentences(text):
    sentences = 0

    for i in text:
        if i == '.' or i == '?' or i == '!':
            sentences += 1

    return sentences


# Counts how many words are in the given text
def count_words(text):
    words = 0

    for i in text:
        if i == " ":
            words += 1
    words += 1

    return words


# Counts how many letters are in the given text
def count_letters(text):
    letters = 0

    for i in text:
        if i.isalpha():
            letters += 1

    return letters


# Prompts the user for input string
def get_text():
    text = get_string("Text: ")
    return text


# Call the main function
if __name__ == '__main__':
    main()

File: week2/problem set/problem set 2/readability/readability.c
#include <cs50.h>
#include <stdio.h>
#include <string.h>
#include <math.h>

// Function declarations
int count_letters(string text);
int count_words(string text);
int count_sentences(string text);
float calculate_index(int letters, int words, int sentences);

int main(void)
{
    // Prompts the user for the text
    string text = get_string("Text: ");

    // These integer variables store the amount of letters, words and sentences in the text given by the user
    int letters = count_letters(text);
    int words = count_words(text);
    int sentences = count_sentences(text);
    // This integer variable stores the level of grade
    int grade = calculate_index(letters, words, sentences);
    // Types out the level of grade
    if (grade < 1)
    {
        // Prints if grade is less than 1
        printf("Before Grade 1\n");
    }
    else if (grade > 16)
    {
        // Prints if grade is greater than 16
        printf("Grade 16+\n");
    }
    else
    {
        // Prints if grade is between 1 and 16
        printf("Grade %i\n", grade);
    }
}

// This function calculates grade level of text
// according to Coleman-Liau index and returns the rounded value
float calculate_index(int letters, int words, int sentences)
{
    // Calculates letters per words
    double L = (float) letters / words * 100;
    // Calculates words per sentences
    double S = (float) sentences / words * 100;
    // Coleman-Liau index calculation
    float index = 0.0588 * L - 0.296 * S - 15.8;
    // Stores rounded value of the index
    int rounded_index = round(index);
    return rounded_index;
}

// This function counts sentences in the given text
int count_sentences(string text)
{
    // Stores the length of given text
    int length = strlen(text);
    // This variable is used as sentence counter
    int sentences = 0;

    for (int i = 0; i < length; i++)
    {
        // Checks if character is period, exclamation point or question mark
        if (text[i] == 33 || text[i] == 46 || text[i] == 63)
        {
            // Adds 1 to sentence counter
            sentences++;
        }
    }
    return sentences;
}

// This function counts words in the given text
int count_words(string text)
{
    // Stores the length of given text
    int length = strlen(text);
    // This variable is used as word counter
    int words = 0;

    for (int i = 0; i < length; i++)
    {
        // Space between character counts as word
        if (text[i] == 32)
        {
            // Adds one to words counter
            words++;
        }
        // If the last character is period, exclamation point
        // or question mark the last word is counted as well
        if (i == length - 1 && (text[i] == 33 || text[i] == 46 || text[i] == 63))
        {
            words++;
        }
    }
    return words;
}

// This function counts letters in the given text
int count_letters(string text)
{
    // Stores the length of given text
    int length = strlen(text);
    // This variable is used as letter counter
    int letters = 0;

    for (int i = 0; i < length; i++)
    {
        // Uppercase and lowercase letters count as one
        if ((text[i] > 64 && text[i] < 91) || (text[i] > 96 && text[i] < 123))
        {
            letters++;
        }
    }
    return letters;
}